File: parent/986 -- Interval List Intersections/tutorial2.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        # Uses the Sweep Line Algorithm
        # Time: O((m + n)log(m + n))
        # Space: O(m + n)
        events = []

        for l, r in firstList:
            events.append((l, 1))
            events.append((r, -1))
        for l, r in secondList:
            events.append((l, 1))
            events.append((r, -1))
        
        events.sort(key=lambda x: (x[0], -x[1]))
        
        res = []
        active = 0
        new_start = 0

        for pos, typ in events:
            if typ == 1:
                active += 1
                if active == 2:
                    new_start = pos
            else:
                if active == 2:
                    res.append([new_start, pos])
                active -= 1
        
        return res

File: parent/986 -- Interval List Intersections/tutorial2.go
package main

import "sort"

func intervalIntersection(firstList [][]int, secondList [][]int) [][]int {
	events := [][2]int{}

	for _, entry := range firstList {
		events = append(events, [2]int{entry[0], 1})
		events = append(events, [2]int{entry[1], -1})
	}
	for _, entry := range secondList {
		events = append(events, [2]int{entry[0], 1})
		events = append(events, [2]int{entry[1], -1})
	}

	sort.Slice(events, func(i, j int) bool {
		if events[i][0] != events[j][0] {
			return events[i][0] < events[j][0]
		}
		return events[i][1] > events[j][1]
	})

	res := [][]int{}
	new_start := 0
	active := 0

	for _, entry := range events {
		if entry[1] == 1 {
			active++
			if active == 2 {
				new_start = entry[0]
			}
		} else {
			if active == 2 {
				res = append(res, []int{new_start, entry[0]})
			}
			active--
		}
	}

	return res
}
